### PR TITLE
address some bandit security recommendations

### PR DESCRIPTION
**Pull request by MartinRyan**
quoting subprocess commands,  jinja2 set autoescaping to true. cherrpy wsgiserver3 socket permissions set to 755.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -48,3 +48,25 @@ passphrase
...
 coverage.xml
 htmlcov/
+
+\.vscode/settings\.json
+
+\.pytest_cache/v/cache/
+
+\.vscode/launch\.json
+
+lib/cherrypy/test/access\.log
```
Comment (ColemanTom) [inline .gitignore]: Consider instead something like `lib/cherrypy/**/*.log` if you are wanting all of those log files to be ignored, including in nested directories

File: .gitignore [hunk 2]
```diff
@@ -48,3 +48,25 @@ passphrase
 .coverage.*
 coverage.xml
 htmlcov/
+
+\.vscode/settings\.json
```
Comment (ColemanTom) [inline .gitignore]: just ignore `.vscode` instead of two different files

File: lib/cylc/subprocess_safe.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+""" Function to sanitize input to a spawning subprocess where shell==True
+    Bandit B602: subprocess_popen_with_shell_equals_true
+    https://docs.openstack.org/developer/bandit/plugins/subprocess_popen_with_shell_equals_true.html
+    REASON IGNORED:
+    Bandit can't determine if the command input is sanitized, it just raises
+    an issue if it detects Popen with with the option shell=True and so nosec
```
Comment (kinow) [inline lib/cylc/subprocess_safe.py]: s/with with/with

File: lib/cylc/subprocess_safe.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    is used here to supress false positives.
```
Comment (kinow) [inline lib/cylc/subprocess_safe.py]: s/supress/suppress/

File: .gitignore [hunk 5]
```diff
@@ -48,3 +48,25 @@ passphrase
...
+
+lib/cherrypy/test/access\.log
+
+lib/cherrypy/test/error\.log
+
+lib/cherrypy/test/test\.access\.log
+
+lib/cherrypy/test/test\.error\.log
+
+Pipfile
```
Comment (ColemanTom) [inline .gitignore]: It looks like this gitignore includes comments above sections describing what they are, it would be better to follow that example I think. Also you shouldn't need to escape `.`

File: lib/cherrypy/test/test_xmlrpc.py [hunk 6]
```diff
@@ -1,13 +1,21 @@
...
 try:
-    from xmlrpclib import DateTime, Fault, ProtocolError, ServerProxy
-    from xmlrpclib import SafeTransport
+    from xmlrpclib import (DateTime,  # nosec
+                           Fault,
+                           ProtocolError,
+                           ServerProxy,
+                           SafeTransport)
+    # from defusedxml.xmlrpclib import DateTime, Fault, ProtocolError 
+    # from defusedxml.xmlrpclib import ServerProxy, SafeTransport
```
Comment (ColemanTom) [inline lib/cherrypy/test/test_xmlrpc.py]: Is this a WIP? You have commented out imports here, which are referred to by your link above.

File: lib/cherrypy/wsgiserver/wsgiserver3.py [hunk 7]
```diff
@@ -1627,9 +1627,15 @@ def start(self):
...
+            # So everyone can access the socket, this was set to 777
+            # https://docs.openstack.org/developer/bandit/plugins/set_bad_file_permissions.html
+            # This plugin test looks for the use of chmod and will alert when
+            # it is used to set particularly permissive control flags.
+            # A MEDIUM warning is generated if a file is set to group
+            # executable and a HIGH warning is reported if a file
+            # is set world writable. Warnings are given with HIGH confidence.
             try:
-                os.chmod(self.bind_addr, 0o777)
+                os.chmod(self.bind_addr, 0o755)
```
Comment (kinow) [inline lib/cherrypy/wsgiserver/wsgiserver3.py]: Agree it's safer to reduce permission, but note this is a change to CherryPy. Might be a good idea to send an issue/comment to the upstream repository, and see what they think about it.

Note that in their current code, there is a `TODO` note, suggesting that they could be accepting a pull request to make that configurable https://github.com/cherrypy/cheroot/blob/6c0e8c896116f65b51678b61acd22931f6cbdf84/cheroot/server.py#L1810, but I think we won't put any efforts into that, given we are moving away from CherryPy.

Also, there is a small chance of issues in some sites, depending on the operating system and permissions for user/folder. From http://man7.org/linux/man-pages/man7/unix.7.html, under "Pathname socket ownership and permissions"

>       In the Linux implementation, pathname sockets honor the permissions
>       of the directory they are in.  Creation of a new socket fails if the
>       process does not have write and search (execute) permission on the
>       directory in which the socket is created.
>
>       On Linux, connecting to a stream socket object requires write permis‐
>       sion on that socket; sending a datagram to a datagram socket likewise
>       requires write permission on that socket.  POSIX does not make any
>       statement about the effect of the permissions on a socket file, and
>       on some systems (e.g., older BSDs), the socket permissions are
>       ignored.  Portable programs should not rely on this feature for secu‐
>       rity.

So in theory, something like 640 or 660 could work too, but there's no guarantee going from `777` to `755` or more restrictive won't affect users. So in case it gets merged, we may need to include something some where, like in the release notes I think.

**Comment by kinow** [comment]
Added a few comments @MartinRyan , but it wasn't a review. Just a couple of possible typos, and a note about the socket permission, as I think we may need to add a note about it in release notes if it gets included in a next release. 👏 great work chasing all the issues around the code

File: .gitignore (hunk 1 above)
Comment (MartinRyan) [reply .gitignore]: sounds good Tom

File: .gitignore (hunk 2 above)
Comment (MartinRyan) [reply .gitignore]: ;-)

File: lib/cherrypy/test/test_xmlrpc.py (hunk 6 above)
Comment (MartinRyan) [reply lib/cherrypy/test/test_xmlrpc.py]: this was unintentional, from a wip

File: lib/cherrypy/wsgiserver/wsgiserver3.py (hunk 7 above)
Comment (MartinRyan) [reply lib/cherrypy/wsgiserver/wsgiserver3.py]: Thanks Bruno

File: lib/parsec/jinja2support.py [hunk 8]
```diff
@@ -95,7 +95,9 @@ def jinja2environment(dir_=None):
     env = Environment(
         loader=ChoiceLoader([FileSystemLoader(dir_), PyModuleLoader()]),
         undefined=StrictUndefined,
-        extensions=['jinja2.ext.do'])
+        extensions=['jinja2.ext.do'],
+        autoescape=True
```
Comment (TomekTrzeciak) [inline lib/parsec/jinja2support.py]: Probably not the setting you want enabled for the suite templating:
```
{% set task_chain = 'foo => bar => baz' %}
{% autoescape True %}
graph = {{task_chain}}
{% endautoescape %}
```
outputs:
```
graph = foo =&gt; bar =&gt; baz
```

See also this in regard to enabling it globally by default: http://jinja.pocoo.org/docs/2.10/faq/#why-is-autoescaping-not-the-default

File: lib/cherrypy/test/modfastcgi.py [hunk 9]
```diff
@@ -38,14 +38,15 @@
...
+from cylc.subprocess_safe import popencylc
 
 import cherrypy
 from cherrypy.process import plugins, servers
 from cherrypy.test import helper
 
 
 def read_process(cmd, args=""):
-    pipein, pipeout = os.popen4("%s %s" % (cmd, args))
+    pipein, pipeout = popencylc("%s %s" % cmd, args)
```
Comment (matthewrmshin) [inline lib/cherrypy/test/modfastcgi.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cherrypy/test/modfcgid.py [hunk 10]
```diff
@@ -46,7 +47,7 @@
 
 
 def read_process(cmd, args=""):
-    pipein, pipeout = os.popen4("%s %s" % (cmd, args))
+    pipein, pipeout = popencylc("%s %s" % cmd, args)
```
Comment (matthewrmshin) [inline lib/cherrypy/test/modfcgid.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cherrypy/test/modpy.py [hunk 11]
```diff
@@ -38,12 +38,13 @@
...
 import re
 import time
+from cylc.subprocess_safe import popencylc
 
 from cherrypy.test import helper
 
 
 def read_process(cmd, args=""):
-    pipein, pipeout = os.popen4("%s %s" % (cmd, args))
+    pipein, pipeout = popencylc("%s %s" % cmd, args)
```
Comment (matthewrmshin) [inline lib/cherrypy/test/modpy.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cherrypy/test/modwsgi.py [hunk 12]
```diff
@@ -37,13 +37,14 @@
...
 import time
+from cylc.subprocess_safe import popencylc
 
 import cherrypy
 from cherrypy.test import helper, webtest
 
 
 def read_process(cmd, args=""):
-    pipein, pipeout = os.popen4("%s %s" % (cmd, args))
+    pipein, pipeout = popencylc("%s %s" % cmd, args)
```
Comment (matthewrmshin) [inline lib/cherrypy/test/modwsgi.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cherrypy/test/test_xmlrpc.py [hunk 13]
```diff
@@ -1,13 +1,19 @@
...
+                           ServerProxy,
+                           SafeTransport)
 except ImportError:
-    from xmlrpc.client import DateTime, Fault, ProtocolError, ServerProxy
-    from xmlrpc.client import SafeTransport
+    from xmlrpc.client import (DateTime,
+                               Fault,
+                               ProtocolError,
+                               ServerProxy,
+                               SafeTransport)
```
Comment (matthewrmshin) [inline lib/cherrypy/test/test_xmlrpc.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cherrypy/wsgiserver/wsgiserver3.py (hunk 7 above)
Comment (matthewrmshin) [reply lib/cherrypy/wsgiserver/wsgiserver3.py]: We should probably delete this source file. Cylc does not use this.

File: lib/cylc/subprocess_safe.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+    try:
+        command = quote(cmd)
+        process = Popen(command, bufsize, executable, stdin, stdout,  # nosec
+                        stderr, preexec_fn, close_fds, shell, cwd, env,
+                        universal_newlines, startupinfo, creationflags)
+        return process
+    except OSError as exc:
+        LOG.exception(exc)
+        sys.exit(str(exc))
```
Comment (matthewrmshin) [inline lib/cylc/subprocess_safe.py]: Should just re-raise the exception?

File: tests/cylc-cat-state/basic/state-check.py [hunk 15]
```diff
@@ -30,8 +31,9 @@ def main(argv):
     sname = argv[0]
     rundir = argv[1]
 
-    p = subprocess.Popen("cylc cat-state " + sname, shell=True, 
-                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    p = popencylc("cylc cat-state " + sname,
+                  shell=True, stdout=subprocess.PIPE,
+                  stderr=subprocess.PIPE)
```
Comment (matthewrmshin) [inline tests/cylc-cat-state/basic/state-check.py]: Command can change to `['cylc', 'cat-state', sname]` so we can get rid of the `shell=True` argument.

File: lib/cherrypy/wsgiserver/wsgiserver3.py (hunk 7 above)
Comment (matthewrmshin) [reply lib/cherrypy/wsgiserver/wsgiserver3.py]: Note: the bundled version of CherryPy is already heavily modified, so removing more unused-unsafe source files is more than acceptable.

**Comment by matthewrmshin** [comment]
Looks like Codacy is failing because it is *clever* enough to pass on the security issue from subprocess to the new wrapping module?

**Comment by MartinRyan** [comment]
yes matt, perhaps there's a codacy tag similar to # nosec that might help?

File: lib/parsec/jinja2support.py [hunk 16]
```diff
@@ -93,9 +94,14 @@ def jinja2environment(dir_=None):
         dir_ = os.getcwd()
 
     env = Environment(
+        autoescape=select_autoescape(
+            enabled_extensions=('rc'),
```
[the remark this answers is not in the document]
Comment (MartinRyan) [reply lib/parsec/jinja2support.py]: thanks Tomek

File: lib/parsec/jinja2support.py [hunk 17]
```diff
@@ -93,9 +94,14 @@ def jinja2environment(dir_=None):
         dir_ = os.getcwd()
 
     env = Environment(
+        autoescape=select_autoescape(
+            enabled_extensions=(),
+            default_for_string=False,
+            default=True),
```
Comment (TomekTrzeciak) [inline lib/parsec/jinja2support.py]: Sorry, but this change still does enable autoescape for all files (the default value should be `False`).
```suggestion
            default=False),
```

Comment (matthewrmshin) [reply lib/parsec/jinja2support.py]: In the context of this logic here, we only use Jinja2 to format local files e.g. `suite.rc`, so there is no untrusted tainted network stuff going on, and so there is no security issue at all.

**Comment by hjoliver** [comment]
@MartinRyan - to save me reading everything about, can you summarize the state of this PR?  Is it ready to go as far you're concerned?

**Comment by hjoliver** [comment]
@MartinRyan - this branch is now conflicted (one file only).

Should be an easy fix, looks like you added one line to the LaTeX User Guide, which has just been deleted and replaced with the Sphinx version.

File: bin/cylc-get-host-metrics [hunk 18]
```diff
@@ -35,19 +35,18 @@ keys as requested via the OPTIONS:
...
 from subprocess import Popen, PIPE, CalledProcessError
 import json
 
 from cylc.option_parsers import OptionParser
+import sys
+if '--use-ssh' in sys.argv[1:]:
+    sys.argv.remove('--use-ssh')
+    from cylc.remote import remrun
+    if remrun():
+        sys.exit(0)
```
Comment (hjoliver) [inline bin/cylc-get-host-metrics]: @MartinRyan - can you explain why you made this change?  Import order is a style issue, not security, isn't it?  And the same ordering is in many other files in bin/ that you haven't changed. The existing order is just a minor efficiency thing - no point in importing stuff that will never be used if the command just re-invokes itself on a remote host (which is what `remrun` does).

Comment (MartinRyan) [reply bin/cylc-get-host-metrics]: codacy was compalining and failing failing the build hilary

Comment (hjoliver) [reply bin/cylc-get-host-metrics]: Fair enough!  Codacy "fails" if the number of flagged issues goes up, but there's a whole bunch we need we need to deal with (potentially via the Codacy equivalent of a `#nosec` - as in this case, I believe - as these static analysis tools blindly apply rules with no understanding of context or reasonable exceptions).

**Comment by hjoliver** [comment]
Looks like you've accidentally committed a bunch of built (non-source) files from the old User Guide, @MartinRyan. The old (LaTeX) User Guide has been nuked on master.
